src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java

```diff
@@ -0,0 +1,268 @@
+                .whenScenarioStateIs("1")
+                .willSetStateTo("2")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withStatus(404)
+                        .withStatusMessage("404 Not Found")
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\"message\": \"File not found\"}")));
+
+        // Force an error to test fallback refreshing from agent
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("2")
+                .willSetStateTo("3")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withStatus(404)
+                        .withStatusMessage("404 Not Found")
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\"message\": \"File not found\"}")));
+
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("3")
+                .willSetStateTo("4")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../AppCredentials/files/body-githubapp-create-installation-accesstokens.json")));
+
+    }
+
+    @Test
+    public void testAgentRefresh() throws Exception {
+        long notStaleSeconds = GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS;
+        try {
+            appCredentials.setApiUri(githubApi.baseUrl());
+
+            // We want to demonstrate successful caching without waiting for a the default 1 minute
+            // Must set this to a large enough number to avoid flaky test
+            GitHubAppCredentials.AppInstallationToken.NOT_STALE_MINIMUM_SECONDS = 15;
+
+            final String jenkinsfile = String.join(
+                "\n",
+                "// run checkout several times",
+                "node ('my-agent') {",
+                // First Checkout on agent should use cached token passed via remoting
```

Comment:
```suggestion
                "  echo 'First Checkout on agent should use cached token passed via remoting'",
```

etc. so that `BuildWatcher` can show comments in context?